File: 백준/Silver/1189. 컴백홈/컴백홈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    private static int R;
    private static int C;
    private static int K;
    private static int[] resultCnt;
    private static boolean[][] visited;

    private static boolean[][] canNotGo;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] temp = br.readLine().split(" ");
        R = Integer.parseInt(temp[0]);
        C = Integer.parseInt(temp[1]);
        K = Integer.parseInt(temp[2]);
        resultCnt = new int[R*C+1];
        visited = new boolean[R][C];
        canNotGo = new boolean[R][C];
        for (int i = 0 ; i < R ; i++) {
            temp = br.readLine().split("");
            for (int j = 0; j < C; j++) {
                if (temp[j].equals("T")) canNotGo[i][j] = true;
            }
        }
        visited[R-1][0] = true;
        dfs(0, R-1, 0);
        System.out.println(resultCnt[K]);
    }
    private static void dfs(int x, int y, int cnt) {
        cnt++;
        if (x == C-1 && y == 0) {
            resultCnt[cnt]++;
            return;
        }
        int[] moveX = {+1, -1, 0, 0};
        int[] moveY = {0, 0, +1, -1};
        for (int i = 0 ; i < 4 ; i++) {
            int toX = x + moveX[i];
            int toY = y + moveY[i];
            if (toX < 0 || toY < 0 || toX >= C || toY >= R) continue;
            if (!visited[toY][toX] && !canNotGo[toY][toX]) {
                visited[toY][toX] = true;
                dfs(toX, toY, cnt);
                visited[toY][toX] = false;
            }
        }
    }
}

File: 백준/Silver/1189. 컴백홈/컴백홈.py
def solution():
    R, C, K = map(int, input().split())
    matrix = [list(map(str, input())) for _ in range(R)]
    cnt = [0]
    def dfs(x, y, depth):
        if x == 0 and y == C-1 and depth == K:
            cnt[0] += 1
            return
        if depth >= K:
            return
        
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]
            if 0 > nx or 0 > ny or nx >= R or ny >= C: continue
            if vis[nx][ny] == 0 and matrix[nx][ny] == '.':
                vis[nx][ny] = 1
                dfs(nx, ny, depth + 1)
                vis[nx][ny] = 0

    vis = [[0 for _ in range(C)] for _ in range(R)]
    vis[R-1][0] = 1
    dfs(R-1, 0, 1)
    print(cnt[0])

solution()